### PR TITLE
Zarr: handle numpy.datetime64/timedelta64 extension types in V3

### LINKED ISSUE #13912: Zarr V3: support `numpy.datetime64` extension data type
### Feature description

GDAL's Zarr driver does not support reading Zarr V3 arrays that use the `numpy.datetime64` extension data type. This is a registered Zarr V3 extension: https://github.com/zarr-developers/zarr-extensions/tree/main/data-types/numpy.datetime64

Related: #13782 (string extension data types)

PR incoming to support the extension (will leave the indexing failure for another PR). 

### Additional context

Support for datetime64 in the Zarr driver was noted as out of scope in the original driver PR (#3896):

> datetime64 ("M") and timedelta64 ("m") data types aren't supported (we could likely implement them as double if needed)

For Zarr V2 this was rarely a problem in practice because xarray writes time coordinates using CF-convention encoding (int64 + `units` attribute), so GDAL never encounters the raw numpy dtype. In Zarr V3, zarr-python 3.x writes time coordinates directly as the `numpy.datetime64` extension data type by default. This means every Zarr V3 store with a time coordinate produced by xarray or zarr-python 3.x will contain this extension type.

### 1. Dimension name and coordinate value loss (all arrays)

When a data array references a `numpy.datetime64` coordinate via `_ARRAY_DIMENSIONS`, GDAL falls back to anonymous dimension names. Band metadata shows `DIM_dim0_INDEX=0` instead of `DIM_time_INDEX=0` and time coordinate values are absent from band metadata entirely.

### 2. Failure for 4D+ dimension-indexed access

For arrays with 2 or more extra dimensions (e.g. time × z × lat × lon), the dimension index syntax (`:0,0`) fails:

```
ERROR 1: Wrong number of indices of extra dimensions
```

GDAL cannot resolve the `time` dimension, which appears to cause a miscount of extra dimensions. The `:0,0` indices are rejected even though there genuinely are 2 extra dimensions (time and z).

3D arrays (1 extra dimension) degrade more gracefully — `:0` works, just with anonymous dimension names.

### 3. Time coordinate array completely inaccessible

Opening the time array directly fails:

```
ERROR 1: Invalid or unsupported format for data_type: { "name": "numpy.datetime64", "configuration": { "unit": "ns", "scale_factor": 1 } }
```

## Reprex

A self-contained bash script creates a minimal local Zarr V3 store using zarr-python 3.x and exercises all access patterns. 

The script is at: https://gist.github.com/mdsumner/b526797a561002902833aa8bf5291eda

The store contains:
- `time` coordinate: 3 timestamps as `numpy.datetime64[ns]`
- `z` coordinate: 2 depths as `float64`
- `lat`, `lon` coordinates: `float64`
- `sst`: Float32, shape (time=3, lat=4, lon=5) — 1 extra dim
- `chla`: Float32, shape (time=3, z=2, lat=4, lon=5) — 2 extra dims

### Results summary

| Test | 3D sst | 4D chla |
|------|--------|---------|
| Bare array (extra dims → bands) | ✓ degraded dims | ✓ degraded dims |
| Dimension index `:0` / `:0,0` | ✓ degraded dims | **FAIL** |
| `gdal_translate` slice to GeoTIFF | ✓ | **FAIL** |
| VRT band selection | ✓ | ✓ |

Other results:
- `gdalinfo` on group: succeeds with errors on stderr, subdatasets listed
- `gdalmdiminfo` on group: succeeds with errors on stderr, dimensions anonymous
- `gdalinfo` on time array: **FAIL**
- Coordinate arrays (lat, lon, z): all succeed

The 4D failure matches what is observed on a real-world cloud dataset (NOAA/NWFSC CHLA-Z, public Zarr V3 on GCS):

```bash
export GS_NO_SIGN_REQUEST=YES
# FAILS: "Wrong number of indices of extra dimensions"
gdalinfo 'ZARR:"/vsigs/nmfs_odp_nwfsc/CB/fish-pace-datasets/chla-z/zarr":/CHLA:0,0'
```

## The extension data type

The `time` array's `zarr.json` contains:

```json
{
  "data_type": {
    "name": "numpy.datetime64",
    "configuration": {
      "unit": "ns",
      "scale_factor": 1
    }
  }
}
```

### PR DESCRIPTION
**Pull request by mdsumner**
Zarr V3: support numpy.datetime64 extension data type, fixes missing extension support of #13912

Adds an else if branch in ParseDtypeV3 that handles JSON object extension types with "name": "numpy.datetime64" or "numpy.timedelta64", mapping them to GDT_Int64. 

Adds three test cases to `autotest/gdrivers/zarr_driver.py`:

- test_zarr_read_numpy_datetime64_extension_zarr_v3 — Parametrized over both numpy.datetime64 and numpy.timedelta64. Creates a synthetic V3 store with the extension type (no fallback key, matching what zarr-python 3.x writes). Asserts it opens and maps to GDT_Int64.

- test_zarr_read_numpy_datetime64_unsupported_extension_zarr_v3 — Verifies that an unrecognized extension type (without fallback) correctly raises an error.

## What are related issues/pull requests?

Fixes #13912 and relates to similar extension enhancement request #13912

## Tasklist

 - [x] AI (Claude, chat and agent) supported my development of this PR
 - [x] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [x] Add test case(s)
 - [ ] Add documentation
- [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: autotest/gdrivers/zarr_driver.py [hunk 1]
```diff
@@ -2731,6 +2731,59 @@ def test_zarr_read_data_type_fallback_zarr_v3(tmp_vsimem):
...
+    """Test that numpy.datetime64 and numpy.timedelta64 extension types
+    written by zarr-python 3.x (without a fallback key) are mapped to Int64."""
+
+    j = {
+        "zarr_format": 3,
+        "node_type": "array",
+        "shape": [3],
+        "data_type": {
+            "name": extension_name,
+            "configuration": {"unit": "ns", "scale_factor": 1},
```
Comment (rouault) [inline autotest/gdrivers/zarr_driver.py:2752]: we should likely expose those configuration attributes as "data_type.unit" and "data_type.scale_factor" attributes so users can make sense of the array values

And looking at https://github.com/zarr-developers/zarr-extensions/tree/main/data-types/numpy.datetime64#fill-value-representation, the "NaT" fill_nodata value should be taken into account

It would also be worth enhancing the test to read actual (non-zero) array values to check that then endianness is correctly handled

File: frmts/zarr/zarr_v3_array.cpp [hunk 2]
```diff
@@ -1202,6 +1202,22 @@ static GDALExtendedDataType ParseDtypeV3(const CPLJSONObject &obj,
...
+        {
+            const auto osName = obj.GetString("name");
+            if (osName == "numpy.datetime64" || osName == "numpy.timedelta64")
+            {
+                DtypeElt elt;
+                elt.nativeType = DtypeElt::NativeType::SIGNED_INT;
+                elt.gdalType = GDALExtendedDataType::Create(GDT_Int64);
+                elt.gdalSize = elt.gdalType.GetSize();
+                elt.nativeSize = elt.gdalSize;
+                if (elt.nativeSize > 1)
```
Comment (rouault) [inline frmts/zarr/zarr_v3_array.cpp]: the if() test can be removed: the condition is always true

File: autotest/gdrivers/zarr_driver.py (hunk 1 above)
Comment (rouault) [reply autotest/gdrivers/zarr_driver.py:2752]: > we should likely expose those configuration attributes as "data_type.unit" and "data_type.scale_factor"

and also "data_type.name"

**Comment by mdsumner** [comment]
Thanks Even!  Addressed those in a second commit. I'm less comfortable with the second part so here is a review summary from the digital friend, which might be interesting in its own right: 


- `NaT` handling (lines 1666-1673) — correctly placed before `ParseNoDataStringAsDouble` which would choke on "NaT". Mapping to `INT64_MIN` matches the extension spec. The `GDT_Int64` guard is appropriate.
- `oOrigDtype` capture (line 1573) — correctly saved before the fallback substitution at line 1574, so the structural info block at line 1832 reads the original extension metadata regardless of whether a fallback key was present.
- Structural info exposure (lines 1832-1856) — clean, defensive (checks validity/non-empty before each SetStructuralInfo), and placed alongside the existing `SetStructuralInfo("COMPRESSOR", ...)` pattern.
- Test — thorough. Real chunk data, round-trip via multidim API, checks structural info, checks NaT nodata, removed the redundant `@gdaltest.enable_exceptions()` from the success test.

To consider: 

The `nativeSize > 1` guard removal at line 1196 is in the string-type branch, not the datetime branch. That affects bool, int8, uint8 — on big-endian systems, those 1-byte types now get `needByteSwapping = true`, which triggers an unnecessary decoded buffer allocation in `NeedDecodedBuffer()`. It's harmless (byte-swapping 1 byte is identity), but it's a drive-by change to unrelated code. The same removal at line 1210 (datetime branch) is fine since `Int64` is always 8 bytes. Reviews might ask why the string branch was touched — I'd either restore the guard on line 1196 or split it into a separate cleanup commit so the datetime PR stays focused. 

The `scale_factor > 0` guard silently drops zero or missing values, which is correct for the spec. The `NaT` check being keyed on `GDT_Int64` rather than specifically the datetime extension name is slightly broad, but any non-datetime `Int64` with fill_value: "NaT" would be a nonsensical zarr.json anyway.

File: frmts/zarr/zarr_v3_array.cpp [hunk 3]
```diff
@@ -1193,14 +1193,24 @@ static GDALExtendedDataType ParseDtypeV3(const CPLJSONObject &obj,
             elt.gdalSize = elt.gdalType.GetSize();
             if (!elt.gdalTypeIsApproxOfNative)
                 elt.nativeSize = elt.gdalSize;
-
-            if (elt.nativeSize > 1)
+            elt.needByteSwapping = (CPL_IS_LSB == 0);
```
Comment (rouault) [inline frmts/zarr/zarr_v3_array.cpp]: The removal of the if test here is incorrect. Curious if it is a (human) misunderstanding in what I wanted (removal for the datetime branch), or if the machine "hallucinated"

**Comment by rouault** [comment]
> so here is a review summary from the digital friend,

per our future policy https://github.com/OSGeo/gdal/pull/13880 this won't be allowed

File: frmts/zarr/zarr_v3_array.cpp (hunk 3 above)
Comment (rouault) [inline frmts/zarr/zarr_v3_array.cpp]:
```suggestion
            if (elt.nativeSize > 1)
            {
                elt.needByteSwapping = (CPL_IS_LSB == 0);
            }
```

**Comment by rouault** [comment]
I've rebased on top of latest master and fix the endianness issue in tests

**Comment by mdsumner** [comment]
Thanks, and apologies that was a misguided step on my part.
